socket-LB: Terminate pod netns connections to deleted service backends

aditighag opened the pull request:
Extension of https://github.com/cilium/cilium/pull/25169 to terminate connections to deleted service backends in pod network namespaces.

Testing:

Manually tested with starting a pair of UDP server and pod applications, and deleted the service backend.

```
$ k get pods -o wide
NAME          READY   STATUS    RESTARTS   AGE    IP             NODE          NOMINATED NODE   READINESS GATES
client        1/1     Running   0          163m   10.244.1.228   kind-worker   <none>           <none>

$ kubectl exec -it server -- iperf3 -s -p  5010
-----------------------------------------------------------
Server listening on 5010 (test #1)
-----------------------------------------------------------
Accepted connection from 10.244.1.228, port 42620
[  5] local 10.244.1.204 port 5010 connected to 10.244.1.228 port 35150

```

```
time="2024-06-28T20:53:54Z" level=debug msg="handling udp connections to deleted backend 10.244.1.204:5010" subsys=service
time="2024-06-28T20:53:54Z" level=info msg="socket {35150 5010 10.244.1.228 10.244.1.204 0 [297562409 0]}" subsys=datapath-sockets
time="2024-06-28T20:53:54Z" level=debug msg="Destroyed socket: {35150 5010 10.244.1.228 10.244.1.204 0 [297562409 0]}" subsys=datapath-sockets
time="2024-06-28T20:53:54Z" level=info msg="Forcefully terminated sockets" errors="<nil>" failed=0 filter="{10.244.1.204 5010 2 17 0x4584220}" subsys=datapath-sockets success=1
```

Fixes: https://github.com/cilium/cilium/issues/27300

```release-note
Forcefully terminate stale connections in pod network namespaces that are connected to deleted service backends when socket-lb is enabled, and allow pod applications to re-connect to active backends.
```

aditighag committed:
install: Mount /var/run/netns

This is required to be able to exec into pod
network namespaces by opening `/var/run/netns/<pod-netns>`.
The current use case is to be able to terminate stale
pod connections when socket-Lb is enabled.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

borkmann reviewed (approved):
Looks great! Could you later do a follow-up with adding an explicit test case as well? We do have lots of coverage for services, but I'm thinking UDP connect app specifically.

ysksuzuki commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2276,18 +2278,18 @@ envoy:
-        labelSelector:
-          matchLabels:
-            k8s-app: cilium
+        - topologyKey: kubernetes.io/hostname
```
Is this change irrelevant to the terminatePodConnections flag?

aditighag replied:
Err... I don't see these unrelated changes in my local git branch. 😕 

Edit: I'll revert the changes. 🤦‍♀️

ysksuzuki commented on pkg/service/connections.go:
```diff
@@ -71,10 +77,54 @@ func (s *Service) TerminateUDPConnectionsToBackend(l3n4Addr *lb.L3n4Addr) {
+				continue
+			}
+			err = ns.Do(func() error {
+				return s.backendConnectionHandler.Destroy(sockets.SocketFilter{
```
Question: If there are netns that have nothing to do with pods, we iterate those netns too, but we don't destroy unrelated sockets, right? Because we check cilium_lb[4,6]_reverse_sk before destroying sockets.

borkmann replied:
iiuc, this should also take the socket cookie into account so we don't destroy anything unrelated

aditighag replied:
>If there are netns that have nothing to do with pods, we iterate those netns too, 

AFAIK, non-pod network namespaces are not mounted by container runtimes under the path we are iterating. 

>but we don't destroy unrelated sockets, right? Because we check cilium_lb[4,6]_reverse_sk before destroying sockets.

Yes, that's right.

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1016,6 +1016,8 @@ socketLB:
   # -- Disable socket lb for non-root ns. This is used to enable Istio routing rules.
   # hostNamespaceOnly: false
+  # -- Enable terminating pod connections to deleted service backends.
+  # terminatePodConnections: true
```
Commenting this out will mean it's undocumented in the Helm reference (due to a bug in helm-docs, c.f. https://github.com/cilium/cilium/pull/20639#issuecomment-1197895181). If you want this field to be documented, you likely need to uncomment it with a default value

aditighag replied:
>Commenting this out will mean it's undocumented in the Helm reference

Yes, that's intentional.

joestringer replied:
Can we set it with `terminatePodConnections: ~` instead? That _should_ provide no overriding default but allow the docs generation to work.

ysksuzuki commented on daemon/cmd/daemon_main.go:
```diff
@@ -345,6 +345,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Bool(option.EnableSocketLB, false, "Enable socket-based LB for E/W traffic")
 	option.BindEnv(vp, option.EnableSocketLB)
 
+	flags.Bool(option.EnableSocketLBPodConnectionTermination, true, "Enable terminating connections to deleted service backends when socket-LB is enabled")
```
Question: Are there any use cases where users would want to disable this flag? Is this for the backward compatibility?

borkmann replied:
I'd think we always want to have this enabled whenever users have KPR enabled. The flag to disable could also be hidden imho and does not necessarily need to be explicitly exposed via helm.

aditighag replied:
>I'd think we always want to have this enabled whenever users have KPR enabled.

+1

I've also added the flag to Helm settings, but it's not documented so it's hidden.

gandro reviewed (commented):
Thanks!

ysksuzuki reviewed (commented):
Nice work! Just some non-blocking comments and questions.

Also, we would want to update the document here
https://github.com/cilium/cilium/blob/main/Documentation/network/kubernetes/kubeproxy-free.rst?plain=1#L1637-L1646

aditighag reviewed (commented):
>Could you later do a follow-up with adding an explicit test case as well?

Yes, that's on the follow-up list.

aditighag committed:
docs: Update socket-LB connection termination note

Update the section as the agent now terminates stale
connections in pod network namespaces as well.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

learnitall reviewed (approved):
Doc changes LGTM, thanks!

aditighag committed:
config: Add flag for socket LB pod connections termination

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Terminate stale pod netns connections

Extension of PR#25169 [1] to terminate pod netns connections
to deleted service backends in pod network namespaces.

[1] https://github.com/cilium/cilium/pull/25169

Signed-off-by: Aditi Ghag <aditi@cilium.io>

ysksuzuki reviewed (approved):
LGTM!

chancez commented on pkg/service/connections.go:
```diff
@@ -71,10 +77,54 @@ func (s *Service) TerminateUDPConnectionsToBackend(l3n4Addr *lb.L3n4Addr) {
+	}
+
+	// Iterate over all pod network namespaces, and terminate any stale connections.
+	if option.Config.EnableSocketLBPodConnectionTermination && !option.Config.BPFSocketLBHostnsOnly {
```
Why does this depend on `BPFSocketLBHostnsOnly`?

aditighag replied:
>Why does this depend on BPFSocketLBHostnsOnly?

`BPFSocketLBHostnsOnly`: Socket-LB is enabled only for the host netns.

chancez replied:
And what about hostNS pods that have stale UDP connections? Wouldn't we still want to terminate those?

gandro commented on pkg/service/connections.go:
```diff
@@ -71,10 +77,54 @@ func (s *Service) TerminateUDPConnectionsToBackend(l3n4Addr *lb.L3n4Addr) {
+		}
+
+		for _, file := range files {
+			ns, err := netns.OpenPinned(filepath.Join(defaults.NetNsPath, file.Name()))
```
I just noticed this very late: Does this require `CAP_SYS_ADMIN`?

We've been trying to not require that (and on modern kernel it should no longer be required). If we're reintroducing high privileged permissions to cilium-agent, we ought to document that here

https://github.com/cilium/cilium/blob/e0fc5e6a0ca3bf80d060a836fd28783d88f10dea/install/kubernetes/cilium/values.yaml#L265-L268

aditighag replied:
Thanks for highlighting this. I removed it to see is something breaks, and this is where it fails - 

`time="2024-07-15T14:23:35Z" level=fatal msg="failed to probe helper" error="detect support for FnGetNetnsCookie for program type CGroupSock: detect support for CGroupSock: load program: operation not permitted" helper=FnGetNetnsCookie programtype=CGroupSock subsys=probes`

This helper is called here in bpf_sock which has been around since the early days - https://github.com/cilium/cilium/blob/main/bpf/bpf_sock.c#L80. Looks like the [note](https://github.com/cilium/cilium/blob/e0fc5e6a0ca3bf80d060a836fd28783d88f10dea/install/kubernetes/cilium/values.yaml#L265-L268) needs to be updated?


Edit: I may not have the supported versions of container runtimes on my cluster - 

```
# Both PERFMON and BPF requires kernel 5.8, container runtime
# cri-o >= v1.22.0 or containerd >= v1.5.0.
```

aditighag replied:
Looking at the `setns man page`, entering pod network namespaces would require `CAP_SYS_ADMIN`. 

/cc @borkmann 

Edit: We can document this requirement in the kpr documentation.

borkmann replied:
Doesn't the agent already have this req anyway wrt health endpoint launcher which spawns a new netns and moves devices in there?

aditighag replied:
^^ Same q as Daniel... Will hold off doing any doc changes for now.

joestringer replied:
Let's document it here in the YAML. While I agree it would be good to enumerate the permissions Cilium needs and come up with a path for lower permissions in future, I don't think it's meaningful from the [threat model](https://docs.cilium.io/en/stable/security/threat-model/#threat-model) to only disable `CAP_SYS_ADMIN` and leave some of the other privileges in place.

gandro replied:
Fully agreed with Joe. In general, we need to get better in documenting why the agent needs certain capabilities. I didn't think about health-checking, I think you are 100% correct on this Daniel. But it is also easily turned off and many production users don't run with health-checking enabled for various other reasons (it causes unnecessary traffic, occupies one pod IP which is expensive in ENI mode etc). Similarly for this feature in the PR here, which is not relevant for users who don't use SockLB or KPR.

Some users might be willing to turn of these features if it means they can significantly lower the permissions that cilium-agent has.

aditighag replied:
Absolutely agreed on documenting the requirements for capabilities, was waiting on a confirmation for the health endpoint dep. Pushed the update.

ldelossa reviewed (approved):
LGTM. So w/r/t #25169 we took an approach where we don't necessary join the host mount ns, but just mount in the namespace dir stead?

aditighag committed:
install: Document requirements for SYS_ADMIN

Switching into non-root network namespaces requires
SYS_ADMIN permissions. The health endpoint infrastructure
has had this dependency, socket-LB will have a similar
requirement.

Suggested-by: Daniel Borkmann <daniel@iogearbox.net>
Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Aditi Ghag <aditi@cilium.io>
